### PR TITLE
issue #51 - adding ability to specify explicit dependencies when depl…

### PR DESCRIPTION
**Pull request by bfitzpat**
…oying integration

Signed-off-by: Brian Fitzpatrick <bfitzpat@redhat.com>

### CODE REVIEW

**Comment by bfitzpat** [comment]
image

image

image

image

image

**Comment by bfitzpat** [comment]
There are two things that need to be expanded on for this functionality.
1. Validating the name of a given camel component. Right now we don't have a good way to do that inside JS/TS
2. Validating that a given maven dependency actually exists and is accessible

@lhein or @apupier - any ideas how to improve on these? 

Right now all I'm doing is validating that there are three colons (group:artifact:version) (though "group:artifact:" is valid even though it shouldn't be)... and validating that there's a dash (i.e. camel-component). This is obviously not enough validation, so I'd love some input on how to improve it

**Comment by bfitzpat** [comment]
On the plus side, we should be able to apply the same multi-prompt logic for other types of deployments such as properties https://github.com/camel-tooling/vscode-camelk/pull/88 or resources https://github.com/camel-tooling/vscode-camelk/pull/85

**Comment by apupier** [comment]
> 2\. Validating that a given maven dependency actually exists and is accessible

might used Maven command line to see if a dependency is available or not: https://www.npmjs.com/package/maven

or use https://www.npmjs.com/package/mvn-artifact-download and try download to see if it is available

subsidiary question: how/which extra repositories are configured for Camel K?

**Comment by apupier** [comment]
> 1. Validating the name of a given camel component. Right now we don't have a good way to do that inside JS/TS

you mean a camel-component supported by Camel K? or that the name is a syntaxically valid component name?

**Comment by bfitzpat** [comment]
> 
> 
> > 1. Validating the name of a given camel component. Right now we don't have a good way to do that inside JS/TS
> 
> you mean a camel-component supported by Camel K? or that the name is a syntactically valid component name?

Ideally both -- is it a valid component name and is it downloadable/accessible

**Comment by bfitzpat** [comment]
> > 2. Validating that a given maven dependency actually exists and is accessible
> might used Maven command line to see if a dependency is available or not: https://www.npmjs.com/package/maven
> or use https://www.npmjs.com/package/mvn-artifact-download and try download to see if it is available

I looked at the NPM Maven component and didn't see a way to use it to test a dependency. Maybe mvn-artifact-dependency could work, but I don't see an easy way to try that. Do you have any ideas?

> subsidiary question: how/which extra repositories are configured for Camel K?

Considering the whole Apache Camel K project is written in Go with very little maven, I can't see anywhere they're explicitly calling out particular repositories - but it does seem that they're getting Camel 3 artifacts from somewhere.

**Comment by apupier** [comment]
> > > 1. Validating that a given maven dependency actually exists and is accessible
> > >    might used Maven command line to see if a dependency is available or not: https://www.npmjs.com/package/maven
> > >    or use https://www.npmjs.com/package/mvn-artifact-download and try download to see if it is available
> 
> I looked at the NPM Maven component and didn't see a way to use it to test a dependency. Maybe mvn-artifact-dependency could work, but I don't see an easy way to try that. Do you have any ideas?

I never tried and don't the API but i would expect that at least, when trying to download an artefact if not available, it is throwing an error that can be caught to determine if the artifact is missing (for big artifacts it can cause a delay).
Hope that there are some smarter way to do it in one for the npm packages, needs investigation

**Comment by apupier** [comment]
> > subsidiary question: how/which extra repositories are configured for Camel K?
> 
> Considering the whole Apache Camel K project is written in Go with very little maven, I can't see anywhere they're explicitly calling out particular repositories - but it does seem that they're getting Camel 3 artifacts from somewhere.

Camel 3 artifacts are available on Central Maven
http://repo1.maven.org/maven2/org/apache/camel/camel-core/3.0.0-RC1/

It is possible to configure extra Maven repositories, https://github.com/apache/camel-k#configure-maven-settings . As they don't have a Project structure,  it requires to allow the maven repo configuration from the same command and to connect to Camel K instance to retrieve the already configured repo. This seems a bit too complicated for a first iteration.

**Comment by bfitzpat** [comment]
@apupier seems we just leave it pretty open for now and do more validation in the future when things are a bit more well defined on this front?

**Comment by apupier** [comment]
> seems we just leave it pretty open for now and do more validation in the future when things are a bit more well defined on this front?

it seems a fair amount of work, so I vote for that too. Can I let you open issue(s) with opened questions/identified pain points?

**Comment by bfitzpat** [comment]
@apupier - spun off a new issue https://github.com/camel-tooling/vscode-camelk/issues/117

File: src/IntegrationUtils.ts [hunk 1]
```diff
@@ -197,7 +246,7 @@ function createNewIntegration(integrationFileUri: vscode.Uri, devMode? : boolean
 				console.error(error);
```
Comment (lhein) [inline src/IntegrationUtils.ts]: do we need to log it?

File: src/IntegrationUtils.ts [hunk 2]
```diff
@@ -207,15 +256,18 @@ function createNewIntegration(integrationFileUri: vscode.Uri, devMode? : boolean
...
-			}			
+			});
+		}			
+		if (dependencyArray && dependencyArray.length > 0) {
+			dependencyArray.forEach(dependency => {
+				commandString += ` --dependency=${dependency}`;
+			});
 		}
-		commandString += ' "' + absoluteRoot + '"';
 		console.log(`commandString = ${commandString}`);
```
Comment (lhein) [inline src/IntegrationUtils.ts]: logging required?

Comment (bfitzpat) [reply src/IntegrationUtils.ts]: Yes, I like having it there more for debugging purposes than anything. The user never sees it.

File: src/IntegrationUtils.ts [hunk 3]
```diff
@@ -257,3 +310,19 @@ export async function isCamelKAvailable(): Promise<boolean> {
...
+function validateDependency(text : string) {
+	let inputEmpty : boolean = (text === undefined || text.trim().length === 0);
+	if (inputEmpty) {
+		return 'Dependency must be specified in the form of an Apache Camel component id or a Maven dependency in the form of group:artifact:version';
+	}
+	if (!inputEmpty && text.indexOf(':') > -1) {
+		let trifold : string[] = text.split(':');
+		// make sure we have three colon-delimited strings
+		return !(trifold && trifold.length === 3) ? 'Dependency must be specified in the form of a Maven dependency as group:artifact:version' : null;
+	} else if (!(!inputEmpty && text.indexOf('-') > -1)) {
```
Comment (apupier) [inline src/IntegrationUtils.ts]: if I understand well, it requires a `-` in the component id.
I don't think that it is a requirements. For instance there are ahc, apns, amqp, file, timer, ... Or I'm looking to the wrong thing?

Comment (lhein) [reply src/IntegrationUtils.ts]: its about artifactId I thought....like camel-ftp and so on

Comment (bfitzpat) [reply src/IntegrationUtils.ts]: Lars is right - if you look here - https://camel.apache.org/components/latest/ - and click through to ahc, apns, etc. - the dependency always seems to boil down to camel-component

File: src/IntegrationUtils.ts [hunk 4]
```diff
@@ -257,3 +310,19 @@ export async function isCamelKAvailable(): Promise<boolean> {
...
+		// assume we're a camel component for now
+		return 'Dependency must be specified in the form of an Apache Camel component id or a Maven dependency in the form of group:artifact:version';
```
Comment (lhein) [inline src/IntegrationUtils.ts]: maybe the text is confusing?

File: src/IntegrationUtils.ts (hunk 3 above)
Comment (apupier) [reply src/IntegrationUtils.ts]: ok, then maybe worth precising i tin the message that i tis the "camel component artifact id" and not the "camel component id"

File: src/IntegrationUtils.ts (hunk 4 above)
Comment (bfitzpat) [reply src/IntegrationUtils.ts]: Recommendation? Maybe change to "Apache Camel component Artifact Id"?

Comment (bfitzpat) [reply src/IntegrationUtils.ts]: Updated

**Pull request merged**